File: C11/백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.c
#include <stdio.h>

int main() {
    int N;
    scanf("%d", &N);

    int A[1001];
    for (int i = 0; i < N; i++) {
        scanf("%d", &A[i]);
    }

    int dp_inc[1001] = {0};
    int dp_dec[1001] = {0};

    for (int i = 0; i < N; i++) {
        dp_inc[i] = 1;
        for (int j = 0; j < i; j++) {
            if (A[j] < A[i] && dp_inc[j] + 1 > dp_inc[i]) {
                dp_inc[i] = dp_inc[j] + 1;
            }
        }
    }

    for (int i = N - 1; i >= 0; i--) {
        dp_dec[i] = 1;
        for (int j = N - 1; j > i; j--) {
            if (A[j] < A[i] && dp_dec[j] + 1 > dp_dec[i]) {
                dp_dec[i] = dp_dec[j] + 1;
            }
        }
    }

    int result = 0;
    for (int i = 0; i < N; i++) {
        int len = dp_inc[i] + dp_dec[i] - 1;
        if (len > result) result = len;
    }

    printf("%d\n", result);
    return 0;
}

File: Java/백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[] A = new int[N];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            A[i] = Integer.parseInt(st.nextToken());
        }

        int[] dp_inc = new int[N];
        int[] dp_dec = new int[N];

        for (int i = 0; i < N; i++) {
            dp_inc[i] = 1;
            for (int j = 0; j < i; j++) {
                if (A[j] < A[i] && dp_inc[i] < dp_inc[j] + 1) {
                    dp_inc[i] = dp_inc[j] + 1;
                }
            }
        }

        for (int i = N - 1; i >= 0; i--) {
            dp_dec[i] = 1;
            for (int j = N - 1; j > i; j--) {
                if (A[j] < A[i] && dp_dec[i] < dp_dec[j] + 1) {
                    dp_dec[i] = dp_dec[j] + 1;
                }
            }
        }

        int result = 0;
        for (int i = 0; i < N; i++) {
            result = Math.max(result, dp_inc[i] + dp_dec[i] - 1);
        }

        System.out.println(result);
    }
}